// stryker-api/src/config/Config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("../../core");
var Config = /** @class */ (function () {
    function Config() {
        this.mutate = [
            '{src,lib}/**/*.js?(x)',
            '!{src,lib}/**/__tests__/**/*.js?(x)',
            '!{src,lib}/**/?(*.)+(spec|test).js?(x)',
            '!{src,lib}/**/*+(Spec|Test).js?(x)'
        ];
        this.logLevel = core_1.LogLevel.Information;
        this.fileLogLevel = core_1.LogLevel.Off;
        this.timeoutMS = 5000;
        this.timeoutFactor = 1.5;
        this.plugins = ['stryker-*'];
        this.reporter = [];
        this.reporters = ['progress', 'clear-text'];
        this.coverageAnalysis = 'off';
        this.testRunner = 'command';
        this.mutator = 'es5';
        this.transpilers = [];
        this.maxConcurrentTestRunners = Infinity;
        this.symlinkNodeModules = true;
        this.port = undefined;
        this.thresholds = {
            break: null,
            high: 80,
            low: 60
        };
        this.allowConsoleColors = true;
    }
    Config.prototype.set = function (newConfig) {
        var _this = this;
        if (newConfig) {
            Object.keys(newConfig).forEach(function (key) {
                if (typeof newConfig[key] !== 'undefined') {
                    _this[key] = newConfig[key];
                }
            });
        }
    };
    return Config;
}());
exports.default = Config;
